format/FormatMultiImage.py

```diff
@@ -237,10 +237,16 @@ def get_imageset(
                 goniometer = []
                 scan = []
                 for i in range(format_instance.get_num_images()):
-                    beam.append(format_instance.get_beam(i))
-                    detector.append(format_instance.get_detector(i))
-                    goniometer.append(format_instance.get_goniometer(i))
-                    scan.append(format_instance.get_scan(i))
+                    if (single_file_indices is None) or (i in single_file_indices):
```

Comment:
On a practical note - in here if we have an imageset with e.g. 10,000 images and we're loading all of them, as a consequence of 

```
            single_file_indices=list(range(*array_range)),
```

this could involve 10,000 tests of whether `i` is in the list, when it is guaranteed to be in there, so the improvement in load time for one image out of 10,000 could be offset substantially by this search (I don't know offhand how slow / fast `i in list` is... ah, not free - see below)

```
Grey-Area i19-1-normal-PdTEET :) $ python
Python 3.6.9 | packaged by conda-forge | (default, Mar  6 2020, 19:19:02) 
[GCC Clang 9.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> l = list(range(10000))
>>> import time
>>> def go():
...   t0 = time.time()
...   for j in range(10000):
...     if j in l:
...       a = 1
...     else:
...       a = 0
...   t1 = time.time()
...   print(t1 - t0)
... 
>>> go()
0.48654699325561523
>>> go()
0.4670860767364502
```

So it's potentially an expensive route - could be worth testing if `len(single_file_indices) < get_num_images()` and only testing if this is true?

Reply:
Hi Graeme, thanks for looking at this.

About a test, I can certainly see the value of illustrating this in a way that's more concrete, but do we have any tools available for testing execution times? Or maybe load a single image out of a multiimage file and use `profile` to count the calls to `get_beam`? (should be ~1, not num_images)

About your second comment, please see my new commit, which gives a second chance to short-circuit testing `i in single_file_indices` using the check of `len` as you suggested:

```
                num_images = format_instance.get_num_images()
                for i in range(num_images):
                    if (
                        single_file_indices is None
                        or len(single_file_indices) == num_images
                        or i in single_file_indices
                    ):
```

Please let me know your thoughts on a test counting calls to `get_beam`.

Reply:
@dagewa Thanks! I did exactly that.

@graeme-winter The new test makes an imageset from 1 image in a 4-image h5 file. We mock the _beam method of the format class and count the calls. Before this PR there would be one call for every image in the file. Now there are 2 calls, one originating from `FormatHDF5SaclaMPCCD._detector` and the other from `FormatMultiImage.get_imageset`.